os/kernel/wqueue/kwork_slmthread.c

```diff
@@ -0,0 +1,238 @@
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+
+#include <errno.h>
+#include <queue.h>
+#include <debug.h>
+
+#include <tinyara/wqueue.h>
+#include <tinyara/kthread.h>
+#include <tinyara/kmalloc.h>
+#include <tinyara/clock.h>
+#include "sched/sched.h"
+#include "wqueue/wqueue.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Type Declarations
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+/* The state of the kernel mode, high priority softlockup monitor work queue. */
+
+struct hp_wqueue_s g_softlockup_monitor_hpwork;
+
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: softlockup_monitor_work
+ *
+ * Description:
+ *   This work function will be called for pre-defined regular interval.
+ *   It does following operations
+ *   1) It will check the number of context switch operations occured since
+ *      it was called the last time
+ *   2) Whether is there any other thread running except idle thread
+ *   3) Based on above information, it will decide which thread is hogging
+ *      CPU and calls PANIC()
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+static void softlockup_monitor_work(FAR void *arg)
+{
+		/*Check for the thread next to line in ready to tun list */
+	FAR struct tcb_s *rtcb = (FAR struct tcb_s *)g_readytorun.head;
+	FAR struct tcb_s *ntcb = rtcb->flink;
+	struct work_s *test_wq;
+
+	if (context_switch_counter < 2 && ntcb && ntcb->pid != 0)
+	{
+#if (CONFIG_TASK_NAME_SIZE > 0)
+		lldbg("Lockup caused by [%s] and PID is %d\n", ntcb->name, ntcb->pid);
+#endif
+		PANIC();
+	}
+	/* Reset the counter */
+	context_switch_counter = 0;
+
+	/* Queue the worker function again */
+	test_wq = (struct work_s *) malloc(sizeof(struct work_s));
```

Comment:
1. when do you free this allocated memroy?
2. remove a space

Reply:
Removed the malloc usage now. Since slm workqueue needs to be alive as long as system is up, used global variable